enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -700,6 +692,58 @@ func (c *FirecrackerContainer) LoadSnapshot(ctx context.Context, workspaceDirOve
 	if err != nil {
 		return status.WrapError(err, "Failed to initialize firecracker VM exec client")
 	}
+
+	if workspaceDirOverride != "" {
+		// If the snapshot is being loaded with a different workspaceFS
+		// then handle that now.
+		wsImgRelPath, err := c.createWorkspaceImage(ctx, workspaceDirOverride)
+		if err != nil {
+			return err
+		}
+		if err := c.hotSwapWorkspace(ctx, execClient, wsImgRelPath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// createWorkspaceImage creates a new ext4 image from the action working dir
+// and returns the chroot-relative path to the created image.
+func (c *FirecrackerContainer) createWorkspaceImage(ctx context.Context, workspacePath string) (string, error) {
+	ctx, span := tracing.StartSpan(ctx)
+	defer span.End()
+
+	c.workspaceGeneration++
+	relativePath := fmt.Sprintf("%s.gen_%d.ext4", workspaceFSName, c.workspaceGeneration)
+	hostPath := filepath.Join(c.getChroot(), relativePath)
+	sizeBytes, err := disk.DirSize(workspacePath)
+	if err != nil {
+		return "", err
+	}
+	if err := ext4.DirectoryToImage(ctx, workspacePath, hostPath, sizeBytes+c.constants.DiskSlackSpaceMB*1e6); err != nil {
+		return "", err
+	}
+	c.workspaceFSPath = hostPath
+	return relativePath, nil
+}
+
+// hotSwapWorkspace unmounts the workspace drive from a running firecracker
+// container, updates the workspace block device to an ext4 image pointed to
+// by chrootRelativeImagePath, and re-mounts the drive.
+func (c *FirecrackerContainer) hotSwapWorkspace(ctx context.Context, execClient vmxpb.ExecClient, chrootRelativeImagePath string) error {
+	ctx, span := tracing.StartSpan(ctx)
+	defer span.End()
+
+	if _, err := execClient.UnmountWorkspace(ctx, &vmxpb.UnmountWorkspaceRequest{}); err != nil {
+		return status.WrapError(err, "Failed to unmount workspace")
+	}
+	if err := c.machine.UpdateGuestDrive(ctx, workspaceDriveID, chrootRelativeImagePath); err != nil {
+		return status.InternalErrorf("error updating workspace drive attached to snapshot: %s", err)
+	}
+	if _, err := execClient.MountWorkspace(ctx, &vmxpb.MountWorkspaceRequest{}); err != nil {
+		return status.WrapError(err, "Failed to remount workspace after update")
```

Comment:
We are terrible at this, and I plan to audit it everywhere at some point, but:

errors message should begin with a lower case letter (because they can be wrapped by other errors higher up)
log lines can begin with an upper case letter (because they are immediately logged and not wrapped)

Reply:
Makes sense, I'll keep this in mind for the future. Fixed the capitalization in the errors added by this PR.